src/cargo/core/compiler/context/compilation_files.rs

```diff
@@ -12,6 +12,34 @@ use super::{BuildContext, Context, FileFlavor, Kind, Layout, Unit};
 use crate::core::{TargetKind, Workspace};
 use crate::util::{self, CargoResult};
 
+/// The `Metadata` is a hash used to make unique file names for each unit in a build.
+/// For example:
+/// - A project may depend on crate `A` and crate `B`, so the package name must be in the file name.
+/// - Similarly a project may depend on two versions of `A`, so the version must be in the file name.
+/// In general this must include all things that need to be distinguished in different parts of
+/// the same build. This is absolutely required or we override things before
+/// we get chance to use them.
+///
+/// We use a hash because it is an easy way to guarantee
+/// that all the inputs can be converted to a valid path.
+///
+/// This also acts as the main layer of caching provided by Cargo.
+/// For example, we want to cache `cargo build` and `cargo doc` separately, so that running one
+/// does not invalidate the artifacts for the other. We do this by including `profile` in the hash,
```

Comment:
The word `profile` here isn't quite right for the example.  To distinguish build/check/doc/etc, the `CompileMode` is included (`unit.mode`).  The profile is included to track some settings that differ, for example `panic=abort` changes for dependencies based on whether it is for a test or not.  Dependencies can also be built multiple times and they need to be kept separate for example when build script profile overrides are used.  There are a few other reasons, but I think this is partially captured in the comment in the compute function.